### PR TITLE
Implemented API Request Model Validation through FluentValidation

### PR DESCRIPTION
**Pull request by contactvpatel**
https://docs.fluentvalidation.net/en/latest/aspnet.html

### CODE REVIEW

**Comment by contactvpatel** [comment]
@barotbm I also would like to move Middleware folder & Response Model from Demo.Util to Demo.Api project since both are only being used in API project but I was getting some dependency injection error for Middleware through Scrutor scan. Please take a look. Demo.Util should only contains components that are shared in more than one projects such as logging.

File: src/Demo.Api/Validators/ProductApiModelValidator.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using Demo.Api.Models;
+using FluentValidation;
+
+namespace Demo.Api.Validators
+{
+    public class ProductApiModelValidator : AbstractValidator<ProductApiModel>
```
Comment (barotbm) [inline src/Demo.Api/Validators/ProductApiModelValidator.cs:6]: If there are simple validations i.e. Required, will it work if the model itself implement it
[Required]
public string Name {get;set;}

are there any advantages of implementing it as a separate validator?

File: src/Demo.Api/Filters/ValidationFilter.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    public class ValidationFilter : IAsyncActionFilter
+    {
+        private readonly ILogger<ValidationFilter> _logger;
+
+        public ValidationFilter(ILogger<ValidationFilter> logger)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
```
Comment (barotbm) [inline src/Demo.Api/Filters/ValidationFilter.cs:24]: This is a nice approach. I really liked the way you are using filters for various things.

**Comment by contactvpatel** [comment]
@barotbm any update on this PR?

**Comment by barotbm** [comment]
> @barotbm I also would like to move Middleware folder & Response Model from Demo.Util to Demo.Api project since both are only being used in API project but I was getting some dependency injection error for Middleware through Scrutor scan. Please take a look. Demo.Util should only contains components that are shared in more than one projects such as logging.

@contactvpatel - looking into it

File: src/Demo.Api/Validators/ProductApiModelValidator.cs (hunk 1 above)
Comment (contactvpatel) [reply src/Demo.Api/Validators/ProductApiModelValidator.cs:6]: It's simply a choice to use data annotations vs fluent validation. Data Annotation is simple but limited in complex validation whereas Fluent Validation servers all kind of complex conditions.
http://www.differencebetween.net/technology/difference-between-data-annotation-and-fluent-validation/
https://stackoverflow.com/questions/6807472/fluent-validation-vs-data-annotations

**Comment by contactvpatel** [comment]
@barotbm I added validation through Data Annotation example in Category Model. So, team can choose any one option based on project need.

**Pull request merged**